### PR TITLE
 Move Absolute Origin Calculation and Marking Dirty region logic to RNS Shell

### PR DESCRIPTION
**Pull request by munezbn**
Now Absolute origin will be calculated after all the mutation is processed.
Nested Children layout issue will be resolved with this.
Implemented multiple helper functions in Layer Classes from TODO list.
Fixed Folly compilation error on Ubuntu 20

### CODE REVIEW

File: ReactSkia/components/RSkComponent.cpp [hunk 1]
```diff
@@ -52,28 +52,39 @@ void RSkComponent::requiresLayer(const ShadowView &shadowView) {
...
       Rect frame = component_.layoutMetrics.frame;
       SkIRect frameIRect = SkIRect::MakeXYWH(frame.origin.x, frame.origin.y, frame.size.width, frame.size.height);
       if(layer() && layer().get())
-        layer().get()->setFrame(frameIRect);
+        layer_->setFrame(frameIRect);
    }
 
-   if(layer_ && layer_->type() == RnsShell::LAYER_TYPE_PICTURE) {
-     RNS_PROFILE_API_OFF(component_.componentName << " getPicture :", static_cast<RnsShell::PictureLayer*>(layer_.get())->setPicture(getPicture()));
+   if(layer_ && layer_.get()) {
```
Comment (shefayaparwin) [inline ReactSkia/components/RSkComponent.cpp:149]: Do the  Layer invalidation needed for State & EventEmitter update as well

File: rns_shell/compositor/layers/Layer.cpp [hunk 2]
```diff
@@ -52,29 +64,94 @@ void Layer::appendChild(SharedLayer child) {
 }
 
 void Layer::insertChild(SharedLayer child, size_t index) {
-    // TODO
-    // Add sanity checks
-    // Check if child already has a parent, and remove it if so
-    RNS_LOG_TODO("Check if child already has a parent, if so then first remove it");
+    if(!child || !child.get()) {
```
Comment (shefayaparwin) [inline rns_shell/compositor/layers/Layer.cpp:67]: Just like to know. Won't the removechild [remove mount instruction] be called before insertchild, if childs hierarchy changes. Is there a use cases demands to do teh removeChild explicitly or just a sanity Handling

File: rns_shell/compositor/layers/Layer.cpp [hunk 3]
```diff
@@ -11,6 +11,16 @@
 
 namespace RnsShell {
 
+#if USE(RNS_SHELL_PARTIAL_UPDATES)
+static inline void addDamgeRect(PaintContext& context, SkIRect dirtyAbsFrameRect) {
+    // TODO Walk through the dirtyRect vector and
+    // 1. If new dirty rect contains inside any pf existing dirtyRect then ignore
```
Comment (shefayaparwin) [inline rns_shell/compositor/layers/Layer.cpp:17]: "pf" refers PartialFrame ?

File: rns_shell/compositor/layers/Layer.cpp [hunk 4]
```diff
@@ -111,30 +189,40 @@ void Layer::paint(PaintContext& context) {
...
+        if (layer == ancestor)
+            return true;
+    }
+    return false;
+}
+
 void Layer::setParent(Layer* layer) {
-    // TODO add checks
-    RNS_LOG_TODO("Add checks");
+    if(layer && layer->hasAncestor(this)) {
```
Comment (shefayaparwin) [inline rns_shell/compositor/layers/Layer.cpp:202]: Is it a case of circled/looped parent & child connection. Is it a sanity Handling or Can this happen in real scenario ?

File: rns_shell/compositor/layers/Layer.cpp [hunk 5]
```diff
@@ -52,29 +64,94 @@ void Layer::appendChild(SharedLayer child) {
...
+}
+
+void Layer::prePaint(PaintContext& context, bool forceLayout) {
+    //Adjust absolute Layout frame and dirty rects
+    preRoll(context, forceLayout);
+    invalidateMask_ = LayerInvalidateNone;
+
+    // prePaint children recursively
+    for (auto& layer : children()) {
+        layer->prePaint(context, (forceLayout || (invalidateMask_ & LayerLayoutInvalidate)));
```
Comment (shefayaparwin) [inline rns_shell/compositor/layers/Layer.cpp]: With invalidateMask_ = LayerInvalidateNone;, will this case satisfy "(invalidateMask_ & LayerLayoutInvalidate)"

File: ReactSkia/components/RSkComponent.cpp (hunk 1 above)
Comment (munezbn) [reply ReactSkia/components/RSkComponent.cpp:149]: No. This needs to be updated now once update props change is integrated

File: rns_shell/compositor/layers/Layer.cpp (hunk 2 above)
Comment (munezbn) [reply rns_shell/compositor/layers/Layer.cpp:67]: In React Native it is true that there is no such scenario ( as far as we know today). But  Rns Shell is written as a generic Shell for Skia, hence it is added as sanity handling.

File: rns_shell/compositor/layers/Layer.cpp (hunk 4 above)
Comment (munezbn) [reply rns_shell/compositor/layers/Layer.cpp:202]: It is just a sanity handling. Like I said in previous comment. RNS shell is written as generic shell for Skia.

File: rns_shell/compositor/layers/Layer.cpp (hunk 3 above)
Comment (munezbn) [reply rns_shell/compositor/layers/Layer.cpp:17]: Typo :). *of

File: rns_shell/compositor/layers/Layer.cpp (hunk 5 above)
Comment (munezbn) [reply rns_shell/compositor/layers/Layer.cpp]: Good catch, Will change

**Review by shefayaparwin — APPROVED** [review]

**Review by pprabhu-ng — APPROVED** [review]

**Pull request merged**